File: 146/LRU_Cache.c
struct LRUCache {
    int key;
    int value;
    struct LRUCache *prev;
    struct LRUCache *next;
};

typedef struct LRUCache LRUCache;

LRUCache *lru_head = NULL;

void move2head(LRUCache *cur) {
    if (lru_head == cur) {
        lru_head = lru_head->next;
        return;
    }

    cur->prev->next = cur->next;
    cur->next->prev = cur->prev;

    cur->next = lru_head;
    cur->prev = lru_head->prev;
    lru_head->prev->next = cur;
    lru_head->prev = cur;

    return ;
}

LRUCache* lRUCacheCreate(int capacity) {
    if (capacity <= 0) return NULL;

    lru_head = (LRUCache *)malloc(sizeof(LRUCache));
    if (!lru_head) return NULL;

    lru_head->key = -1;
    lru_head->value = -1;
    LRUCache *cur = lru_head;

    while(capacity > 1) {
        cur->next = (LRUCache *)malloc(sizeof(LRUCache));
        if (!cur->next) return NULL;
        cur->next->key = -1;
        cur->next->value = -1;

        cur->next->prev = cur;
        cur = cur->next;

        capacity--;
    }
    cur->next = lru_head;
    lru_head->prev = cur;

    return lru_head;
}

int lRUCacheGet(LRUCache* obj, int key) {
    if (!obj) return -1;

    LRUCache *cur = lru_head;
    do {
        if (cur->key == key) {
            move2head(cur);

            return cur->value;
        }
        cur = cur->next;
    } while (cur != lru_head);

    return -1;
}

void lRUCachePut(LRUCache* obj, int key, int value) {
    if (!obj) return;

    LRUCache *cur = lru_head;
    do {
        if (cur->key == key) {
            cur->value = value;
            move2head(cur);

            return;
        }
        cur = cur->next;
    } while (cur != lru_head);

    lru_head->key = key;
    lru_head->value = value;
    lru_head = lru_head->next;

    return;
}

void lRUCacheFree(LRUCache* obj) {
    if (!obj) return;

    LRUCache *cur, *tmp = obj;

    do {
        cur = tmp;
        tmp = tmp->next;
        free(cur);
    } while (tmp && tmp != obj);
    obj = NULL;
    return;
}

/**
 * Your LRUCache struct will be instantiated and called as such:
 * struct LRUCache* obj = lRUCacheCreate(capacity);
 * int param_1 = lRUCacheGet(obj, key);
 * lRUCachePut(obj, key, value);
 * lRUCacheFree(obj);
 */

File: 146/LRU_Cache.py
class Node:
    def __init__(self, key: int, value: int):
        self.key = key
        self.value = value
        self.prev = None
        self.next = None


class LRUCache:

    def __init__(self, capacity: int):
        self.head = Node(-1, -1)
        self.map = {}
        cur = self.head
        for x in range(capacity - 1):
            cur.next = Node(-1, -1)
            cur.next.prev = cur
            cur = cur.next

        cur.next = self.head
        self.head.prev = cur

    def move2Head(self, cur: Node) -> None:
        if self.head == cur:
            self.head = self.head.next
        else:
            cur.prev.next = cur.next
            cur.next.prev = cur.prev

            cur.next = self.head
            cur.prev = self.head.prev
            self.head.prev = cur
            cur.prev.next = cur

    def get(self, key: int) -> int:
        if key in self.map:
            cur = self.map.get(key)
            self.move2Head(cur)
            return cur.value
        else:
            return -1

    def put(self, key: int, value: int) -> None:
        if key in self.map:
            cur = self.map.get(key)
            cur.value = value
            self.move2Head(cur)
        else:
            if self.head.key != -1:
                self.map.pop(self.head.key)
            self.head.key = key
            self.head.value = value
            self.map[key] = self.head
            self.head = self.head.next
        print(self.head.value)


# Your LRUCache object will be instantiated and called as such:
# obj = LRUCache(capacity)
# param_1 = obj.get(key)
# obj.put(key,value)
